### PR TITLE
Optimize deferred credits fetch

### PR DESCRIPTION
**Pull request by sydhds**
* [ ] document all added functions
* [ ] try in sandbox /simulation/labnet
  * [ ] if part of node-launch, checked using the `resync_check` flag
* [ ] unit tests on the added/changed features
  * [ ] make tests compile
  * [ ] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

### CODE REVIEW

File: massa-execution-worker/src/speculative_roll_state.rs [hunk 1]
```diff
@@ -298,51 +298,104 @@ impl SpeculativeRollState {
...
+                println!("def creds len: {}", deferred_credits.credits.len());
+                let mut inserted = 0;
+                for (slot, addr_amount) in deferred_credits.credits {
                     if let Some(amount) = addr_amount.get(address) {
-                        res.entry(*slot).or_insert(*amount);
+                        res.entry(slot).or_insert(*amount);
+                        inserted += 1;
                     };
                 }
+                println!("inserted: {}", inserted);
```
Comment (damip) [inline massa-execution-worker/src/speculative_roll_state.rs:350]: remove print

File: massa-execution-worker/src/speculative_roll_state.rs [hunk 2]
```diff
@@ -298,51 +298,104 @@ impl SpeculativeRollState {
...
+                }
+            }
+
+            // get values from final state
+            {
+                let final_state = self.final_state.read();
+                let deferred_credits = final_state
+                    .get_pos_state()
+                    .get_deferred_credits_range(min_slot.., Some(*address));
+                println!("def creds len: {}", deferred_credits.credits.len());
```
Comment (damip) [inline massa-execution-worker/src/speculative_roll_state.rs:389]: remove print

File: massa-execution-worker/src/speculative_roll_state.rs [hunk 3]
```diff
@@ -298,51 +298,104 @@ impl SpeculativeRollState {
...
+                let mut inserted = 0;
+                for (slot, addr_amount) in deferred_credits.credits {
+                    if let Some(amount) = addr_amount.get(address) {
+                        res.entry(slot).or_insert(*amount);
+                        inserted += 1;
+                    };
+                }
+                println!("inserted: {}", inserted);
```
Comment (damip) [inline massa-execution-worker/src/speculative_roll_state.rs:397]: remove print

File: massa-execution-worker/src/speculative_roll_state.rs (hunk 1 above)
Comment (sydhds) [reply massa-execution-worker/src/speculative_roll_state.rs]: Removed

File: massa-execution-worker/src/speculative_roll_state.rs (hunk 2 above)
Comment (sydhds) [reply massa-execution-worker/src/speculative_roll_state.rs]: Removed

File: massa-api/src/public.rs [hunk 4]
```diff
@@ -832,8 +832,25 @@ impl MassaRpcServer for API<Public> {
                 .collect()
         };
 
+        // Compute a limit (as a slot) for deferred credits as it can be quite huge
+        let bound_ts = MassaTime::now()
+            .checked_add(MassaTime::from_millis(
+                60 * 24 * 60 * 60 * 1000, // 60 Days, 24 hours, 60 minutes, 60 seconds, 1000 milliseconds
```
Comment (damip) [inline massa-api/src/public.rs:838]: no magic numbers, let's make it an API config file parameter

File: massa-execution-worker/src/speculative_roll_state.rs [hunk 5]
```diff
@@ -298,51 +298,96 @@ impl SpeculativeRollState {
...
+                let final_state = self.final_state.read();
+                let deferred_credits = final_state
+                    .get_pos_state()
+                    .get_deferred_credits_range(min_slot.., Some(*address));
+                for (slot, addr_amount) in deferred_credits.credits {
+                    if let Some(amount) = addr_amount.get(address) {
+                        res.entry(slot).or_insert(*amount);
+                    };
+                }
+            }
```
Comment (Leo-Besancon) [inline massa-execution-worker/src/speculative_roll_state.rs:390]: Maybe I'm missing something, but why can't you factor this code with a:
```rust
let range = if let Some(max_slot) = max_slot_ {
  RangeBound(Included(min_slot), Excluded(max_slot))
} else {
  RangeBound(Included(min_slot), Unbounded))
}
```

File: massa-api/src/public.rs (hunk 4 above)
Comment (sydhds) [reply massa-api/src/public.rs]: Done

File: massa-models/src/config/constants.rs [hunk 6]
```diff
@@ -342,6 +342,15 @@ pub const ROLL_COUNT_TO_SLASH_ON_DENUNCIATION: u64 = 1;
 /// Maximum size of executed denunciations
 pub const MAX_DENUNCIATION_CHANGES_LENGTH: u64 = 1_000;
 
+//
+// Constants for Api
+//
+
+/// Delta to compute upper bounds when fetching deferred credits
+pub const DEFERRED_CREDITS_DELTA: MassaTime = MassaTime::from_millis(
```
Comment (damip) [inline massa-models/src/config/constants.rs:350]: this should be in the config file, as it is user-tunable

File: massa-execution-worker/src/speculative_roll_state.rs (hunk 5 above)
Comment (sydhds) [reply massa-execution-worker/src/speculative_roll_state.rs:390]: RangeBounds is a trait and not a type (https://doc.rust-lang.org/std/ops/trait.RangeBounds.html) no?

Comment (Leo-Besancon) [reply massa-execution-worker/src/speculative_roll_state.rs:390]: Haven't tested it myself so I trust you :)

File: massa-models/src/config/constants.rs (hunk 6 above)
Comment (sydhds) [reply massa-models/src/config/constants.rs]: Done

File: massa-api/src/tests/mock.rs [hunk 7]
```diff
@@ -7,7 +7,7 @@ use std::{collections::HashMap, net::SocketAddr};
 use massa_api_exports::config::APIConfig;
 use massa_consensus_exports::{ConsensusBroadcasts, MockConsensusController};
 use massa_execution_exports::{GasCosts, MockExecutionController};
-use massa_models::config::CHAINID;
+use massa_models::config::{CHAINID, DEFERRED_CREDITS_DELTA};
```
Comment (Leo-Besancon) [inline massa-api/src/tests/mock.rs:10]: Remove

File: massa-api/src/tests/mock.rs [hunk 8]
```diff
@@ -67,6 +67,7 @@ pub(crate) fn get_apiv2_server(addr: &SocketAddr) -> (API<ApiV2>, APIConfig) {
         periods_per_cycle: PERIODS_PER_CYCLE,
         last_start_period: 0,
         chain_id: *CHAINID,
+        deferred_credits_delta: DEFERRED_CREDITS_DELTA,
```
Comment (Leo-Besancon) [inline massa-api/src/tests/mock.rs:70]: Hardcode test data

File: massa-execution-worker/src/speculative_roll_state.rs (hunk 5 above)
Comment (Leo-Besancon) [reply massa-execution-worker/src/speculative_roll_state.rs:390]: Seems to be just Bound to have an object apparently :) https://doc.rust-lang.org/std/ops/enum.Bound.html

**Review by Leo-Besancon — APPROVED** [review]
LGTM

**Comment by damip** [comment]
TODO MISSING: GRPC API !

**Review by damip — CHANGES_REQUESTED** [review]
TODO: GRPC API

**Comment by damip** [comment]
Actually let's assume GRPC is OK for now. We'll respond accordingly if there is any issue. Also, given the fact that we will reduce the whole deferred credits file, it should not be an issue.

**Pull request merged**